doc/design/record_fetcher_proposal.md

```diff
@@ -0,0 +1,167 @@
+can use to start that conversation
+* must not alter the manifest or its signature before committing to durable storage
+  * without the signature, we can't detect tampering or associate the manifest with a fetcher
+* should probably losslessly compress most things (we get 40% savings with gzip on the sample manifest below)
+    
+## Design Ideas
+### Document Storage
+First, let's talk document storage. Have record fetchers crawl however they want and upload documents to a document
+storage endpoint. This should ideally be some kind of content-addressed blob storage service. We don't need a lot of
+structure for this, as any relationship between documents will be encoded by the manifests -- the service just needs to
+be able to store the raw documents for later lookup. Backing this storage API with a fairly flat filesystem is fine
+here, we're rarely going to list files, only write and read. If we do need to list files, it'll be for some cleanup
+operation and should not be on any request path.
+
+A big benefit of content-addressed storage here is that we get deduping almost for free -- we may need to zero out
+content that changes on every page load (e.g. "This page requested at HH:MM"), but that's more of an optimization than a
+hard requrement. These savings would apply to both documents which remain unchanged between fetches and documents that
+are shared between records.
+
+The actual document collection service will take a file upload and respond with a simple payload denoting the URI that
+may be used to access the file.
+
+### Manifest Construction
+
+Manifests need to be signed, so we'll just use [RFC 7519](https://tools.ietf.org/html/rfc7519) JWT blobs, encoding the
+manifest's information in the JWT payload. As an example, the payload could resemble:
+
+```json
+{
+  "version": 1,
+  "uuid": "6e566397-6803-49ec-9280-62c00b368e3f",
+  "record_type": "USA/FL/Bay/Court",
+  "case_id": "20000001",
+  "collected_at_ms": 1592618026000,
+  "fetcher_id": "github.com/Police-Data-Accessibility-Project/Scrapers:USA/FL/Bay/Court/scraper/Scraper.py@1db20ce1e63768363d532d46a641722986f63524",
+  "documents": [
+    {
+      "label": "some pdf we'll have to ocr",
+      "content_type": "application/pdf",
+      "digest": "...",
+      "uri": "https://..."
+    }, {
+      "label": "party information",
+      "content_type": "text/html",
+      "digest": "...",
+      "uri": "https://..."
+    }
+  ]
+}	
+```
+
+Determining the PDAP-controlled URIs of the documents will require uploading them first, and then constructing the
+manifest.
+
+### Garbage Collection
+Since this design requires upload documents before writing out a manifest, it's possible that we'll upload docs and then
+fail (or neglect) to upload a manifest. This results in orphaned documents, which are just wasted storage. 
+
+To mitigate this, a periodic (daily? weekly?) GC job will run in order to clean up orphaned documents that have existed
```

Comment:
This is going to grow to be a rather heavy task as the repository grows; can the service receiving the docs buffer them in such a way that a missing manifest will result in the pending docs being purged before being written to the data store? As described garbage collection resource consumption will grow at best linearly as the manifest and document count increase; even when there's nothing to clean up.

Reply:
Added a different approach that requires a bit more coordination, but doesn't require us to enumerate potentially millions of documents (assuming this thing takes off). PTAL.